### PR TITLE
betterproto: support `Struct` and `Value`

### LINKED ISSUE #332: Struct, Value well know type not supported
https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/struct

These should map to `dict[str, any]` and `any`

### PR DESCRIPTION
**Pull request by woodruffw**
## Summary

Fixes #332. See also #549.

~~Opening as a draft until I add tests.~~

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
    - [x] This change has an associated test.
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [x] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by woodruffw** [comment]
NB: I haven't made any documentation changes, since I believe this behavior was intended (in terms of Protobuf's specified behavior) but just buggy before. However, if there's somewhere I should note this (like the CHANGELOG?) let me know and I'm more than happy to add it 🙂 

Similarly, in terms of semver, I believe this is considered a nonbreaking change (since it's a bugfix).

File: src/betterproto/__init__.py [hunk 1]
```diff
@@ -1404,6 +1403,15 @@ def to_dict(
...
             The JSON serializable dict representation of this object.
         """
+        # Mirror of from_dict: Struct's `fields` member is transparently
+        # dispatched through instead.
+        if isinstance(self, Struct):
+            output = {**self.fields}
+            for k in self.fields:
+                if hasattr(self.fields[k], "to_dict"):
+                    output[k] = self.fields[k].to_dict(casing, include_default_values)
+            return output
```
Comment (woodruffw) [inline src/betterproto/__init__.py:1413]: This special-casing is arguably pretty ugly, but I couldn't think of a better (clean) way to accomplish this: `Struct` *is* kind of special in protobuf, in that it's a transparent mapping type (when expressed as JSON) where no other types really are. As such, it doesn't really fit into the `to_dict`/`from_dict` behavior below.

**Comment by Gobot1234** [comment]
Honestly I think just putting this behaviour on the type itself would be the best solution. The generated file has been pretty stable for a long time so I don't really have a problem just tacking this on at this point.

**Comment by woodruffw** [comment]
> Thanks for this! Honestly I think just putting this behaviour on the type itself would be the best solution. The generated file has been pretty stable for a long time so I don't really have a problem just tacking this on at this point.

No problem! Making sure I understand: by "on the type itself," you mean specialize `to_dict()` and `from_dict()` on `Struct` itself?

If so, makes sense -- I can do that in a moment.

**Comment by Gobot1234** [comment]
> > Thanks for this! Honestly I think just putting this behaviour on the type itself would be the best solution. The generated file has been pretty stable for a long time so I don't really have a problem just tacking this on at this point.
> 
> No problem! Making sure I understand: by "on the type itself," you mean specialize `to_dict()` and `from_dict()` on `Struct` itself?
> 
> If so, makes sense -- I can do that in a moment.

Yep exactly

**Comment by woodruffw** [comment]
Alright, moved the specialization to `Struct` itself!

**Comment by woodruffw** [comment]
Gentle ping; please let me know if there's anything else I can do here!

File: src/betterproto/__init__.py [hunk 2]
```diff
@@ -1582,6 +1582,7 @@ def _from_dict_init(cls, mapping: Mapping[str, Any]) -> Mapping[str, Any]:
                     )
 
             init_kwargs[field_name] = value
+
```
Comment (Gobot1234) [inline src/betterproto/__init__.py:1585]:
```suggestion

```

**Review by Gobot1234 — DISMISSED** [review]

**Review by Gobot1234 — APPROVED** [review]

**Comment by Gobot1234** [comment]
Test failures seem unrelated

**Comment by woodruffw** [comment]
> Test failures seem unrelated

Yeah, I think so -- looks like the failure is in `Message._betterproto`, which then fails to generate `cls._betterproto_meta`. But I can't see anything in these changes that would cause that.

**Comment by woodruffw** [comment]
Triaging a bit more: `_get_field_default_gen` has an `issubclass` check that fails because `t` is not a `type` object.

This suggests that something about type hints is hinky, since `t` is resolved from them. Maybe a deferred/non-deferred annotations thing, causing `t` to be a `str` instead?

File: src/betterproto/lib/google/protobuf/__init__.py [hunk 3]
```diff
@@ -2,14 +2,21 @@
 # sources: google/protobuf/any.proto, google/protobuf/api.proto, google/protobuf/descriptor.proto, google/protobuf/duration.proto, google/protobuf/empty.proto, google/protobuf/field_mask.proto, google/protobuf/source_context.proto, google/protobuf/struct.proto, google/protobuf/timestamp.proto, google/protobuf/type.proto, google/protobuf/wrappers.proto
 # plugin: python-betterproto
 # This file has been @generated
+
+from __future__ import annotations
```
Comment (woodruffw) [inline src/betterproto/lib/google/protobuf/__init__.py:6]: Ah yep, this is probably doing it.

**Comment by woodruffw** [comment]
(Also, if I can make an additional request: a new beta release after this lands would be greatly appreciated!)

**Comment by Gobot1234** [comment]
Can't do a beta release till https://github.com/grpc/grpc/pull/35329 is merged and https://github.com/danielgtaylor/python-betterproto/pull/540

**Review by Gobot1234 — APPROVED** [review]
Thanks

**Comment by atomicmac** [comment]
@woodruffw @Gobot1234 Is there anything special that needs to be done for serialization/deserialization of the Structs? From what I'm seeing, instantiating a Struct as exemplified in the tests and then attempting to send it over the wire fails because the Struct's dump method is eventually called, which falls back to the Message.dump method, which treats the struct as a dict, which expects all keys to be of type 'string' and all values to be of type 'message' - if the value is an intrinsic type like str, the catch-all bytes(value) call fails in _preprocess_single because of TypeError('string argument without an encoding')

**Comment by Gobot1234** [comment]
Oh it seems like it might need special casing for dump yeah

**Comment by woodruffw** [comment]
Yep, thanks for catching that @atomicmac, and sorry for the gap there. I can look into a fix tomorrow.

**Comment by woodruffw** [comment]
I've started work on this in #559, but I'm a little stumped on how to actually specialize `dump()` here -- from a quick look `Struct` should be transparently serialized as `Map<string, Value>` in the binary encoding, but I'm not sure what the most idiomatic/straightforward way to do that is.

**Comment by atomicmac** [comment]
Thanks for looking @Gobot1234 & @woodruffw ! Yes @woodruffw I was looking at implementing the specialized dump() and came to the same crossroads. It would be nice to not have to wrap every value in the struct, but then autoconverting to protobuf types becomes involved if you want to keep transmitted sizes as small as possible.

**Comment by woodruffw** [comment]
> Yes @woodruffw I was looking at implementing the specialized dump() and came to the same crossroads. It would be nice to not have to wrap every value in the struct, but then autoconverting to protobuf types becomes involved if you want to keep transmitted sizes as small as possible.

Glad to hear it's not just me 😅 -- I've been thinking about this a bit, and I still don't have a great idea here.

**Pull request merged**